File: service-beacons-node/test/services/version1/BeaconsGrpcServiceV1.test.ts
let assert = require('chai').assert;
let grpc = require('grpc');
let async = require('async');

let services = require('../../../../src/protos/beacons_v1_grpc_pb');
let messages = require('../../../../src/protos/beacons_v1_pb');

import { Descriptor, FilterParams, PagingParams } from 'pip-services3-commons-node';
import { ConfigParams } from 'pip-services3-commons-node';
import { References } from 'pip-services3-commons-node';
import { BeaconTypeV1, BeaconV1 } from '../../../src';
import { BeaconsController } from '../../../src/logic/BeaconsController';
import { BeaconsMemoryPersistence } from '../../../src/persistence/BeaconsMemoryPersistence';
import { BeaconsGrpcConverterV1 } from '../../../src/services/version1/BeaconsGrpcConverterV1';
import { BeaconsGrpcServiceV1 } from '../../../src/services/version1/BeaconsGrpcServiceV1';

var grpcConfig = ConfigParams.fromTuples(
    "connection.protocol", "http",
    "connection.host", "localhost",
    "connection.port", 3000
);

const BEACON1: BeaconV1 = {
    id: '1',
    udi: '00001',
    type: BeaconTypeV1.AltBeacon,
    site_id: '1',
    label: 'TestBeacon1',
    center: { type: 'Point', coordinates: [0, 0] },
    radius: 50
};

const BEACON2: BeaconV1 = {
    id: '2',
    udi: '00002',
    type: BeaconTypeV1.iBeacon,
    site_id: '1',
    label: 'TestBeacon2',
    center: { type: 'Point', coordinates: [2, 2] },
    radius: 70
};

suite('BeaconsV1GrpcServiceV1', () => {

    let service: BeaconsGrpcServiceV1;

    let client: any;

    suiteSetup((done) => {
        let persistence = new BeaconsMemoryPersistence();
        let controller = new BeaconsController();

        service = new BeaconsGrpcServiceV1();
        service.configure(grpcConfig);

        let references: References = References.fromTuples(
            new Descriptor('beacons', 'persistence', 'memory', 'default', '1.0'), persistence,
            new Descriptor('beacons', 'controller', 'default', 'default', '1.0'), controller,
            new Descriptor('beacons', 'service', 'grpc', 'default', '1.0'), service
        );

        controller.setReferences(references);
        service.setReferences(references);

        service.open(null, done);
    });

    suiteTeardown((done) => {
        service.close(null, done);
    });

    setup(() => {
        client = new services.BeaconsV1Client('localhost:3000', grpc.credentials.createInsecure());
    });

    test('CRUD Operations', (done) => {
        let beacon1: BeaconV1;

        async.series([
            
            // Create the first beacon
            (callback) => {
                let request = new messages.BeaconV1ObjectRequest()
                request.setBeacon(BeaconsGrpcConverterV1.fromBeacon(BEACON1))

                client.create_beacon(
                    request,
                    (err, response) => {
                        assert.isNull(err);
                        
                        let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());

                        assert.isObject(beacon);
                        assert.equal(BEACON1.udi, beacon.udi);
                        assert.equal(BEACON1.site_id, beacon.site_id);
                        assert.equal(BEACON1.type, beacon.type);
                        assert.equal(BEACON1.label, beacon.label);
                        assert.isNotNull(beacon.center);

                        callback();
                    }
                );
            },
            // Create the second beacon
            (callback) => {
                let request = new messages.BeaconV1ObjectRequest()
                request.setBeacon(BeaconsGrpcConverterV1.fromBeacon(BEACON2))
                client.create_beacon(
                    request,
                    (err, response) => {
                        assert.isNull(err);
                        let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());

                        assert.isObject(beacon);
                        assert.equal(BEACON2.udi, beacon.udi);
                        assert.equal(BEACON2.site_id, beacon.site_id);
                        assert.equal(BEACON2.type, beacon.type);
                        assert.equal(BEACON2.label, beacon.label);
                        assert.isNotNull(beacon.center);

                        callback();
                    }
                );
            },
            // Get all beacons
            (callback) => {
                let request = new messages.BeaconV1PageRequest();
                
                client.get_beacons(
                    request,
                    (err, response) => {
                        assert.isNull(err);
                        let page = BeaconsGrpcConverterV1.toBeaconsPage(response.getPage());

                        assert.isObject(page);
                        assert.lengthOf(page.data, 2);

                        beacon1 = page.data[0];

                        callback();
                    }
                )
            },
            // Update the beacon
            (callback) => {
                beacon1.label = 'ABC';
                let request = new messages.BeaconV1ObjectRequest()
                request.setBeacon(BeaconsGrpcConverterV1.fromBeacon(beacon1))

                client.update_beacon(
                   request,
                    (err, response) => {
                        assert.isNull(err);
                        let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());

                        assert.isObject(beacon);
                        assert.equal(beacon1.id, beacon.id);
                        assert.equal('ABC', beacon.label);

                        callback();
                    }
                )
            },
            // Get beacon by udi
            (callback) => {
                let request = new messages.BeaconV1UdiRequest()
                request.setBeaconUdi( beacon1.udi)

                client.get_beacon_by_udi(
                    request,
                    (err, response) => {
                        assert.isNull(err);
                        let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());

                        assert.isObject(beacon);
                        assert.equal(beacon1.id, beacon.id);

                        callback();
                    }
                )
            },

            // Calculate position for one beacon
            (callback) => {
                let request = new messages.BeaconV1PositionRequest()
                request.setSiteId('1')
                request.setUdisList(['00001'])

                client.calculate_position(
                    request,
                    (err, response) => {
                        assert.isNull(err);

                        let position = BeaconsGrpcConverterV1.toPosition(response.getPoint());

                        assert.isObject(position);
                        assert.equal('Point', position.type);
                        assert.lengthOf(position.coordinates, 2);
                        assert.equal(0, position.coordinates[0]);
                        assert.equal(0, position.coordinates[1]);

                        callback();
                    }
                )
            },

            // Delete the beacon
            (callback) => {
                let request = new messages.BeaconV1IdRequest()
                request.setBeaconId( beacon1.id)
                client.delete_beacon_by_id(
                   request,
                    (err, response) => {
                        assert.isNull(err);
                        let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());

                        assert.isObject(beacon);
                        assert.equal(beacon1.id, beacon.id);

                        callback();
                    }
                )
            },
            // Try to get deleted beacon
            (callback) => {
                let request = new messages.BeaconV1IdRequest()
                request.setBeaconId( beacon1.id)
                client.get_beacon_by_id(
                    request,
                    (err, response) => {
                        assert.isNull(err);

                        //assert.isEmpty(beacon || null);

                        callback();
                    }
                )
            }

        ], done);
    });
});

File: service-beacons-dotnet/test/Service.Test/Services/Version1/BeaconsGrpcServiceV1Test.cs
using System;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Grpc.Core;
using Nov.MaxSamples.Beacons.Data.Version1;
using Nov.MaxSamples.Beacons.Logic;
using Nov.MaxSamples.Beacons.Persistence;
using PipServices3.Commons.Config;
using PipServices3.Commons.Refer;
using Xunit;
using static BeaconsV1.BeaconsV1;

namespace Nov.MaxSamples.Beacons.Services.Version1
{
    [Collection("Sequential")]
    public class BeaconsGrpcServiceV1Test : IDisposable
    {
        private BeaconV1 BEACON1 = new BeaconV1
        {
            Id = "1",
            Udi = "00001",
            Type = BeaconTypeV1.AltBeacon,
            SiteId = "1",
            Label = "TestBeacon1",
            Center = new CenterObjectV1 { Type = "Point", Coordinates = new double[] { 0, 0 } },
            Radius = 50
        };
        private BeaconV1 BEACON2 = new BeaconV1
        {
            Id = "2",
            Udi = "00002",
            Type = BeaconTypeV1.iBeacon,
            SiteId = "1",
            Label = "TestBeacon2",
            Center = new CenterObjectV1 { Type = "Point", Coordinates = new double[] { 2, 2 } },
            Radius = 70
        };

        private BeaconsMemoryPersistence _persistence;
        private BeaconsController _controller;
        private BeaconsGrpcServiceV1 _service;
        private BeaconsV1Client _client;

        public BeaconsGrpcServiceV1Test()
        {
            _persistence = new BeaconsMemoryPersistence();
            _controller = new BeaconsController();

            var config = ConfigParams.FromTuples(
                "connection.protocol", "http",
                "connection.host", "localhost",
                "connection.port", "3001"
            );

            // this service name is for test only, used by client
            _service = new BeaconsGrpcServiceV1("beacons_v1.BeaconsV1");
            _service.Configure(config);

            var references = References.FromTuples(
                new Descriptor("beacons", "persistence", "memory", "default", "1.0"), _persistence,
                new Descriptor("beacons", "controller", "default", "default", "1.0"), _controller,
                new Descriptor("beacons", "service", "grpc", "default", "1.0"), _service
            );

            _controller.SetReferences(references);

            _service.SetReferences(references);
            _service.OpenAsync(null).Wait();

            Channel channel = new Channel("localhost:3001", ChannelCredentials.Insecure);
            _client = new BeaconsV1Client(channel);
        }

        public void Dispose()
        {
            _service.CloseAsync(null).Wait();
        }

        [Fact]
        public async Task TestCrudOperationsAsync()
        {
            var callOptions = new CallOptions();

            // Create the first beacon
            var reply = await _client.create_beaconAsync(new BeaconsV1.BeaconV1ObjectRequest 
            { 
                CorrelationId = "default",
                Beacon = BeaconsGrpcConverterV1.FromBeacon(BEACON1)
            }, callOptions);

            var beacon = BeaconsGrpcConverterV1.ToBeacon(reply.Beacon);

            Assert.NotNull(beacon);
            Assert.Equal(BEACON1.Udi, beacon.Udi);
            Assert.Equal(BEACON1.SiteId, beacon.SiteId);
            Assert.Equal(BEACON1.Type, beacon.Type);
            Assert.Equal(BEACON1.Label, beacon.Label);
            Assert.NotNull(beacon.Center);

            // Create the second beacon
            reply = await _client.create_beaconAsync(new BeaconsV1.BeaconV1ObjectRequest
            {
                CorrelationId = "default",
                Beacon = BeaconsGrpcConverterV1.FromBeacon(BEACON2)
            }, callOptions);
            
            Assert.Null(reply.Error);
            beacon = BeaconsGrpcConverterV1.ToBeacon(reply.Beacon);

            Assert.NotNull(beacon);
            Assert.Equal(BEACON2.Udi, beacon.Udi);
            Assert.Equal(BEACON2.SiteId, beacon.SiteId);
            Assert.Equal(BEACON2.Type, beacon.Type);
            Assert.Equal(BEACON2.Label, beacon.Label);
            Assert.NotNull(beacon.Center);

            // Get all beacons
            var pageReply = await _client.get_beaconsAsync(new BeaconsV1.BeaconV1PageRequest
            {
                CorrelationId = "default",
                //Filter = empty by default
                Paging = new BeaconsV1.PagingParams
                { 
                    Skip = 0,
                    Take = 100,
                    Total = true
                }
            }, callOptions);

            Assert.Null(reply.Error);
            var page = BeaconsGrpcConverterV1.ToBeaconsPage(pageReply.Page);

            Assert.NotNull(page);
            Assert.Equal(2, page.Data.Count);

            var beacon1 = page.Data[0];

            // Update the beacon
            beacon1.Label = "ABC";

            reply = await _client.update_beaconAsync(new BeaconsV1.BeaconV1ObjectRequest
            {
                CorrelationId = "default",
                Beacon = BeaconsGrpcConverterV1.FromBeacon(beacon1)
            }, callOptions);

            Assert.Null(reply.Error);
            beacon = BeaconsGrpcConverterV1.ToBeacon(reply.Beacon);

            Assert.NotNull(beacon);
            Assert.Equal(beacon1.Id, beacon.Id);
            Assert.Equal("ABC", beacon.Label);

            // Get beacon by udi
            reply = await _client.get_beacon_by_udiAsync(new BeaconsV1.BeaconV1UdiRequest
            {
                CorrelationId = "default",
                BeaconUdi = beacon1.Udi
            }, callOptions);

            Assert.Null(reply.Error);
            beacon = BeaconsGrpcConverterV1.ToBeacon(reply.Beacon);

            Assert.NotNull(beacon);
            Assert.Equal(beacon1.Id, beacon.Id);

            // Delete the beacon
            reply = await _client.delete_beacon_by_idAsync(new BeaconsV1.BeaconV1IdRequest
            {
                CorrelationId = "default",
                BeaconId = beacon1.Id
            }, callOptions);

            Assert.Null(reply.Error);
            beacon = BeaconsGrpcConverterV1.ToBeacon(reply.Beacon);

            Assert.NotNull(beacon);
            Assert.Equal(beacon1.Id, beacon.Id);

            // Try to get deleted beacon
            reply = await _client.get_beacon_by_idAsync(new BeaconsV1.BeaconV1IdRequest
            {
                CorrelationId = "default",
                BeaconId = beacon1.Id
            }, callOptions);

            beacon = BeaconsGrpcConverterV1.ToBeacon(reply.Beacon);
            Assert.Null(beacon);
        }
    }
}
